core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -153,6 +182,7 @@ private[spark] class BarrierCoordinator(
         }
         // Add the requester to array of RPCCallContexts pending for reply.
         requesters += requester
+        allGatherMessages(partitionId) = allGatherMessage
```

Comment:
@jiangxb1987 this addresses your comment about maintaining sorted order in the result